mlflow/gateway/providers/mosaicml.py

```diff
@@ -0,0 +1,226 @@
+        #
+        # {"inputs": [prompt],
+        #  {
+        #    "parameters": {
+        #      "temperature": 0.2
+        #    }
+        #   }
+        # }
+
+        resp = await self._request(
+            self.config.model.name,
+            final_payload,
+        )
+        # Response example (https://docs.mosaicml.com/en/latest/inference.html#text-completion-models)
+        # ```
+        # {
+        #   "outputs": [
+        #     "string",
+        #   ],
+        # }
+        # ```
+        return chat.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "message": {
+                            "role": "assistant",
+                            "content": c,
+                        },
+                        "metadata": {},
+                    }
+                    for c in resp["outputs"]
+                ],
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def completions(self, payload: completions.RequestPayload) -> completions.ResponsePayload:
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+        key_mapping = {
+            "max_tokens": "max_new_tokens",
+        }
+        for k1, k2 in key_mapping.items():
+            if k2 in payload:
+                raise HTTPException(
+                    status_code=400, detail=f"Invalid parameter {k2}. Use {k1} instead."
+                )
+        payload = rename_payload_keys(payload, key_mapping)
+
+        # Handle 'prompt' field in payload
+        prompt = payload.pop("prompt")
+        if isinstance(prompt, str):
+            prompt = [prompt]
+
+        # Construct final payload structure
+        final_payload = {"inputs": prompt, "parameters": payload}
+
+        # Input data structure for Mosaic Text Completion endpoint
+        #
+        # {"inputs": [prompt],
+        #  {
+        #    "parameters": {
+        #      "temperature": 0.2
+        #    }
+        #   }
+        # }
+
+        resp = await self._request(
+            self.config.model.name,
+            final_payload,
+        )
+        # Response example (https://docs.mosaicml.com/en/latest/inference.html#text-completion-models)
+        # ```
+        # {
+        #   "outputs": [
+        #     "string",
+        #   ],
+        # }
+        # ```
+        return completions.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "text": c,
+                        "metadata": {},
+                    }
+                    for c in resp["outputs"]
+                ],
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
```

Comment:
Does `MosaicML` provide toke usage info in the response? If it doesn't, it's fine.

Reply:
It does not 😔